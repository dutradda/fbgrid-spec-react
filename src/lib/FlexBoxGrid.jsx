import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class FlexBoxGrid extends Component {
  render() {
    const gridCells = this.props.spec.cells || [];
    const gridStyle = Object.assign(this.props.spec.style || {}, {display: "flex"});
    const gridCellsElements = [];
    let cellElem;

    for (let i = 0; i < gridCells.length; i++) {
      cellElem = this._renderCell(gridCells, i);
      gridCellsElements.push(cellElem);
    }

    return (
      <div style={gridStyle}>
        {gridCellsElements}
      </div>
    );
  }

  _renderCell(gridCells, i) {
    const cell = gridCells[i];
    let componentElem;

    if (cell.component) {
      componentElem = this._renderCellComponent(cell.component)
    }
    else if (cell.subgrid) {
      componentElem = this._renderCellSubgrid(cell.subgrid)
    }
    else {
      componentElem = null;
    }

    return (
      <div style={cell.style} key={i}>
        {componentElem}
      </div>
    );
  }

  _renderCellComponent(component) {
    let componentCls = this.props.componentsMap[component.name];

    if (componentCls === undefined) {
      componentCls = component.name;
    }

    return React.createElement(
      componentCls,
      component.options
    );
  }

  _renderCellSubgrid(subgrid) {
    return (
      <FlexBoxGrid
        spec={subgrid}
        componentsMap={this.props.componentsMap}
      />
    );
  }

  static propTypes = {
    spec: PropTypes.shape({
      style: PropTypes.object,
      cells: PropTypes.arrayOf(
        PropTypes.oneOfType([
          PropTypes.shape({
            style: PropTypes.object,
            component: PropTypes.object.isRequired
          }),
          PropTypes.shape({
            style: PropTypes.object,
            subgrid: PropTypes.object.isRequired
          })
        ])
      ).isRequired,
      componentsMap: PropTypes.object
    }).isRequired
  };

  static defaultProps = {
    spec: {
      style: {},
      cells: []
    },
    componentsMap: {}
  };
};
