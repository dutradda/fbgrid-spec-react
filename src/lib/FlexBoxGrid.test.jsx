import React from 'react';
import renderer from 'react-test-renderer';
import FlexBoxGrid from './FlexBoxGrid';

const UserTag = props => <p>{props.text}</p>;
const componentsMap = { UserTag };

test('FlexBoxGrid renders built-in tags', () => {
  const spec = {
    style: { flexDirection: 'column' },
    cells: [
      {
        component: { name: 'div' }
      },{
        component: { name: 'div' }
      }
    ]
  };
  const flexGrid = renderer.create(
    <FlexBoxGrid spec={spec} />,
  ).toJSON();

  expect(flexGrid).toMatchSnapshot();
});

test('FlexBoxGrid renders user tags', () => {
  const spec = {
    style: { flexDirection: 'column' },
    cells: [
      {
        component: {
          name: 'UserTag',
          options: { text: 'User Tag 1' }
        }
      },{
        component: {
          name: 'UserTag',
          options: { text: 'User Tag 2' }
        }
      }
    ]
  };
  const flexGrid = renderer.create(
    <FlexBoxGrid
      spec={spec}
      componentsMap={componentsMap}
    />,
  ).toJSON();

  expect(flexGrid).toMatchSnapshot();
});

test('FlexBoxGrid renders subgrid', () => {
  const spec = {
    style: { flexDirection: 'column' },
    cells: [
      {
        component: {
          name: 'UserTag',
          options: {
            text: 'User Tag 1'
          }
        }
      },{
        style: { flexGrow: 1 },
        subgrid: {
          cells: [
            {
              component: {
                name: 'UserTag',
                options: {
                  text: 'User Tag 2'
                }
              }
            },
            {
              component: {
                name: 'UserTag',
                options: {
                  text: 'User Tag 3'
                }
              }
            }
          ]
        }
      }
    ]
  };
  const flexGrid = renderer.create(
    <FlexBoxGrid
      spec={spec}
      componentsMap={componentsMap}
    />,
  ).toJSON();

  expect(flexGrid).toMatchSnapshot();
});

test('FlexBoxGrid renders without cells', () => {
  const spec = {
    style: { flexDirection: 'column' }
  };
  const spy = spyOn(console, 'error');
  const flexGrid = renderer.create(
    <FlexBoxGrid spec={spec} />,
  ).toJSON();

  expect(flexGrid).toMatchSnapshot();
  expect(spy.calls.all()[0].args[0]).toEqual(
    'Warning: Failed prop type: The prop `spec.cells` ' +
    'is marked as required in `FlexBoxGrid`, ' +
    'but its value is `undefined`.\n' +
    '    in FlexBoxGrid (at FlexBoxGrid.test.jsx:105)'
  );
});

test('FlexBoxGrid renders cell without elements', () => {
  const spec = {
    style: { flexDirection: 'column' },
    cells: [ { } ]
  };
  const spy = spyOn(console, 'error');
  const flexGrid = renderer.create(
    <FlexBoxGrid spec={spec} />,
  ).toJSON();

  expect(flexGrid).toMatchSnapshot();
  expect(spy.calls.all()[0].args[0]).toEqual(
    'Warning: Failed prop type: Invalid prop ' +
    '`spec.cells[0]` supplied to `FlexBoxGrid`.\n' +
    '    in FlexBoxGrid (at FlexBoxGrid.test.jsx:124)'
  );
});
