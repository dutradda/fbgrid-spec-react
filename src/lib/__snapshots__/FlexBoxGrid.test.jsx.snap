// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexBoxGrid renders built-in tags 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <div
    style={undefined}
  >
    <div />
  </div>
  <div
    style={undefined}
  >
    <div />
  </div>
</div>
`;

exports[`FlexBoxGrid renders user tags 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <div
    style={undefined}
  >
    <p>
      User Tag 1
    </p>
  </div>
  <div
    style={undefined}
  >
    <p>
      User Tag 2
    </p>
  </div>
</div>
`;

exports[`FlexBoxGrid renders subgrid 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <div
    style={undefined}
  >
    <p>
      User Tag 1
    </p>
  </div>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={undefined}
      >
        <p>
          User Tag 2
        </p>
      </div>
      <div
        style={undefined}
      >
        <p>
          User Tag 3
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`FlexBoxGrid renders without cells 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
/>
`;

exports[`FlexBoxGrid renders cell without elements 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <div
    style={undefined}
  />
</div>
`;
